truffle/src/main/java/org/jruby/truffle/nodes/dispatch/CachedBoxedDispatchNode.java

```diff
@@ -54,7 +65,111 @@ public CachedBoxedDispatchNode(
                 value,
                 method,
                 indirect,
-                dispatchAction);
+                dispatchAction,
+                argumentNodes,
+                isSplatted);
+     }
+     
+    public static RubyNode[] expandedArgumentNodes(RubyContext context,
+            InternalMethod method, RubyNode[] argumentNodes, boolean isSplatted) {
+        final RubyNode[] result;
+
+        boolean shouldExpand = true;
+        if (method == null
+                || method.getSharedMethodInfo().getKeywordArguments() == null) {
+            // no keyword arguments in method definition
+            shouldExpand = false;
+        } else if (argumentNodes.length != 0
+                && !(argumentNodes[argumentNodes.length - 1] instanceof HashLiteralNode)) {
+            // last argument is not a Hash that could be expanded
+            shouldExpand = false;
+        } else if (method.getSharedMethodInfo().getArity() == null
+                || method.getSharedMethodInfo().getArity().getRequired() >= argumentNodes.length) {
+            shouldExpand = false;
+        } else if (isSplatted
+                || method.getSharedMethodInfo().getArity().allowsMore()) {
+            // TODO: make optimization work if splat arguments are involed
+            // the problem is that Markers and keyword args are used when
+            // reading splatted args
+            shouldExpand = false;
+        }
+
+        if (shouldExpand) {
+            List<String> kwargs = method.getSharedMethodInfo().getKeywordArguments();
+           
+            int countArgNodes = argumentNodes.length + kwargs.size() + 1;
+            if (argumentNodes.length == 0) {
+                countArgNodes++;
+            }
+            
+            result = new RubyNode[countArgNodes];
+            int i;
+           
+            for (i = 0; i < argumentNodes.length - 1; ++i) {
+                result[i] = argumentNodes[i];
+            }
+           
+            int firstMarker = i++;
+            result[firstMarker] = new MarkerNode(context, null);
+
+            HashLiteralNode hashNode;
+            if (argumentNodes.length > 0) {
+                hashNode = (HashLiteralNode) argumentNodes[argumentNodes.length - 1];
+            } else {
+                hashNode = HashLiteralNode.create(context, null,
+                        new RubyNode[0]);
+            }
+
+            List<String> restKeywordLabels = new ArrayList<String>();
+            for (int j = 0; j < hashNode.size(); j++) {
+               final String label = ((ObjectLiteralNode) hashNode.getKey(j)).execute(null).toString();
```

Comment:
`ObjectLiteralNode.getObject()` should be used here instead of `execute(null)`.
Will the key always be an ObjectLiteralNode?